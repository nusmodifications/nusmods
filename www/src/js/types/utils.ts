export type Omit<T, K> = Pick<T, Exclude<keyof T, K>>;
