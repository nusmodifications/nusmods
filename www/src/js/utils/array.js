// @flow

import { flatMap } from 'lodash';

/**
 * Mixes the delimiter into the array between each element
 *
 * @param {T[]} array
 * @param {U} delimiter
 * @returns {Array<T|U>}
 */
export function intersperse<T, U>(array: T[], delimiter: U): Array<T | U> {
  return flatMap(array, (item): Array<T | U> => [item, delimiter]).slice(0, -1);
}

export function takeUntil<T>(array: T[], max: number, predicate: (T) => boolean): T[] {
  const filtered = [];

  for (let i = 0; i < array.length && filtered.length < max; i++) {
    if (predicate(array[i])) filtered.push(array[i]);
  }

  return filtered;
}
