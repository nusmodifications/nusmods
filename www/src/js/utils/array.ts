// @flow

import { flatMap } from 'lodash';

/**
 * Mixes the delimiter into the array between each element
 *
 * @param {T[]} array
 * @param {U} delimiter
 * @returns {Array<T|U>}
 */
export function intersperse<T, U>(array: T[], delimiter: U): Array<T | U> {
  return flatMap(array, (item): Array<T | U> => [item, delimiter]).slice(0, -1);
}

export function takeUntil<T>(array: T[], max: number, predicate: (T) => boolean): T[] {
  const filtered = [];

  for (let i = 0; i < array.length && filtered.length < max; i++) {
    if (predicate(array[i])) filtered.push(array[i]);
  }

  return filtered;
}

export function firstNonNull<T>(producers: Array<() => ?T>): ?T {
  for (let i = 0; i < producers.length; i++) {
    const result = producers[i]();
    if (result != null) return result;
  }

  return null;
}
