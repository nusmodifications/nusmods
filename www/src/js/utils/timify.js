// @flow
import type { DayText, Lesson, LessonTime } from 'types/modules';
import {
  getISODay,
  format,
  getHours,
  getMinutes,
  getSeconds,
  setHours,
  setMinutes,
  setSeconds,
  startOfDay,
} from 'date-fns';

// Converts a 24-hour format time string to an index.
// Each index corresponds to one cell of each timetable row.
// Each row may not start from index 0, it depends on the config's starting time.
// 0000 -> 0, 0030 -> 1, 0100 -> 2, ...
export function convertTimeToIndex(time: LessonTime): number {
  const hour = parseInt(time.substring(0, 2), 10);
  const minute = parseInt(time.substring(2), 10);

  // TODO: Expose incorrect offsets to user via UI
  // Currently we round up in half hour blocks, but the actual time is not shown
  let minuteOffset;
  if (minute === 0) {
    minuteOffset = 0;
  } else if (minute <= 30) {
    minuteOffset = 1;
  } else {
    minuteOffset = 2;
  }

  return hour * 2 + minuteOffset;
}

// Reverse of convertTimeToIndex.
// 0 -> 0000, 1 -> 0030, 2 -> 0100, ...
export function convertIndexToTime(index: number): LessonTime {
  const hour: number = parseInt(index / 2, 10);
  const minute: string = index % 2 === 0 ? '00' : '30';
  return (hour < 10 ? `0${hour}` : hour.toString()) + minute;
}

export function formatHour(hour: number): string {
  if (hour === 12) return '12 noon';
  if (hour === 0 || hour === 24) return '12 midnight';
  if (hour < 12) return `${hour}am`;
  return `${hour - 12}pm`;
}

export function getTimeAsDate(time: string | number, date: Date = new Date()): Date {
  const dateNumber = parseInt(time, 10);
  return setHours(setMinutes(startOfDay(date), dateNumber % 100), Math.floor(dateNumber / 100));
}

export function formatTime(time: string | number): string {
  const timeNumber = parseInt(time, 10);

  if (timeNumber === 0) return '12 midnight';
  if (timeNumber === 1200) return '12 noon';

  return format(getTimeAsDate(timeNumber), 'h:mm a').toLowerCase();
}

// Create a new date object with time from the second date object
export function setTime(date: Date, time: Date): Date {
  return setHours(setMinutes(setSeconds(date, getSeconds(time)), getMinutes(time)), getHours(time));
}

export const SCHOOLDAYS: Array<DayText> = [
  'Monday',
  'Tuesday',
  'Wednesday',
  'Thursday',
  'Friday',
  'Saturday',
];
export const DEFAULT_EARLIEST_TIME: LessonTime = '1000';
export const DEFAULT_LATEST_TIME: LessonTime = '1800';

// Given an array of lessons, we calculate the earliest and latest timings based on the lesson timings.
// This bounds will then be used to decide the starting and ending hours of the timetable.
export function calculateBorderTimings(
  lessons: Array<Lesson>,
): { startingIndex: number, endingIndex: number } {
  let earliestTime: number = convertTimeToIndex(DEFAULT_EARLIEST_TIME);
  let latestTime: number = convertTimeToIndex(DEFAULT_LATEST_TIME);
  lessons.forEach((lesson) => {
    earliestTime = Math.min(earliestTime, convertTimeToIndex(lesson.StartTime));
    latestTime = Math.max(latestTime, convertTimeToIndex(lesson.EndTime));
  });
  return {
    startingIndex: earliestTime % 2 === 0 ? earliestTime : earliestTime - 1, // floor to earliest hour.
    endingIndex: latestTime % 2 === 0 ? latestTime : latestTime + 1, // ceil to latest hour.
  };
}

// Gets the current time in hours, 0915 -> 9, 1315 -> 13
export function getCurrentHours(
  now: Date = new Date(), // Used for tests only
): number {
  return now.getHours();
}

// Gets the current time in hours, 0915 -> 15, 1345 -> 45
// Current time to always match Singapore's
export function getCurrentMinutes(
  now: Date = new Date(), // Used for tests only
): number {
  return now.getMinutes();
}

// Monday = 0, Friday = 4, Sunday = 6
export function getDayIndex(date: Date = new Date()): number {
  return getISODay(date) - 1;
}

/**
 * Return a copy of the original Date incremented by the given number of days
 */
export function daysAfter(startDate: Date, days: number): Date {
  const d = new Date(startDate.valueOf());
  d.setUTCDate(d.getUTCDate() + days);
  return d;
}
