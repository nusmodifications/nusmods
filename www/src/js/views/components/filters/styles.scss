@import "~styles/utils/modules-entry.scss";

$v-padding: 0.6rem;
$h-padding: $grid-gutter-width / 2;

.heading {
  margin-bottom: 0;
  font-weight: bold;
  font-size: $font-size-xs;

  @include touchscreen-or-breakpoint-down {
    padding: $v-padding $h-padding;
    font-size: $font-size-sm;
    border-bottom: 1px solid var(--gray-lighter);
  }
}

.label {
  margin: 0;
  font-size: $font-size-xs;

  &,
  label,
  input {
    cursor: pointer;
  }

  &:hover {
    color: theme-color();
  }

  &.enabled {
    font-weight: bold;
    color: theme-color();
  }

  .checklist &,
  .dropdown & {
    display: block;
    padding-top: 0.1rem;
    padding-bottom: 0.1rem;

    @include touchscreen-or-breakpoint-down {
      padding: $v-padding $h-padding $v-padding ($h-padding + 20px);
      border-bottom: 1px solid var(--gray-lighter);
    }
  }

  :global(.text-muted) {
    font-weight: normal;
  }

  @include touchscreen-only {
    font-size: $font-size-sm;
  }
}

.dropdown {
  $icon-size: 1rem;

  h4 label {
    margin-bottom: 0;
  }

  .label {
    white-space: normal;
  }

  .searchWrapper {
    position: relative;
    margin-right: ($grid-gutter-width / 2);

    // To select both the search glass icon and the chevron
    svg {
      position: absolute;
      top: 0.35rem;
      width: $icon-size;
      height: $icon-size;
      cursor: pointer;
      transition: $transition-base;
    }

    .searchIcon {
      left: 0;
    }

    .openIcon {
      right: 0;
    }

    &.focused svg {
      color: theme-color();
    }
  }

  .searchInput {
    padding-left: $icon-size + 0.3rem;
    margin: 0.2rem 0;
    border-width: 0 0 1px;
    border-radius: 0;

    &:focus {
      box-shadow: none;
    }
  }

  :global(.dropdown-menu) {
    width: calc(100% - 0.2rem);
    max-height: 17rem;
    padding: 0;
    margin: 0;
    border-radius: 0 0 3px 3px;

    .label {
      padding: 0;
    }

    :global(.dropdown-selected) .label {
      color: currentColor;
    }
  }

  select {
    width: calc(100% - 3px);
  }

  @include media-breakpoint-down(sm) {
    .heading {
      border-bottom: 0;
    }

    .searchWrapper {
      margin: 0 ($grid-gutter-width / 2);
    }

    select {
      width: calc(100% - #{$grid-gutter-width});
      margin: 0.5rem ($grid-gutter-width / 2);
    }
  }
}

.timeslot {
  margin-bottom: 1rem;

  .label {
    display: block;
    padding: 0.2rem 0;
    margin: 0;
    text-align: center;
    border: 1px solid transparent;

    &.enabled {
      color: #fff;
      background: theme-color();
    }

    &:hover {
      border-color: theme-color();
    }
  }

  @include touchscreen-only {
    .heading {
      border-bottom: 0;
    }

    th {
      font-size: $font-size-sm;
      color: var(--gray-light);
    }

    .label {
      padding: $v-padding 0;
    }
  }
}
