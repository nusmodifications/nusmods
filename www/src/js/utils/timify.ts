import { DayText, LessonTime } from 'types/modules';
import {
  format,
  getHours,
  getISODay,
  getMinutes,
  getSeconds,
  parseISO,
  setHours,
  setMinutes,
  setSeconds,
  startOfDay,
} from 'date-fns';
import { TimePeriod } from 'types/venues';
import { Lesson } from 'types/timetables';

const SGT_OFFSET = -8 * 60;

// Converts a 24-hour format time string to an index.
// Each index corresponds to one cell of each timetable row.
// Each row may not start from index 0, it depends on the config's starting time.
// 0000 -> 0, 0030 -> 1, 0100 -> 2, ...
export function convertTimeToIndex(time: LessonTime): number {
  const hour = parseInt(time.substring(0, 2), 10);
  const minute = parseInt(time.substring(2), 10);

  // TODO: Expose incorrect offsets to user via UI
  // Currently we round up in half hour blocks, but the actual time is not shown
  let minuteOffset;
  if (minute === 0) {
    minuteOffset = 0;
  } else if (minute <= 30) {
    minuteOffset = 1;
  } else {
    minuteOffset = 2;
  }

  return hour * 2 + minuteOffset;
}

// Reverse of convertTimeToIndex.
// 0 -> 0000, 1 -> 0030, 2 -> 0100, ... , 48 -> 2400
export function convertIndexToTime(index: number): LessonTime {
  const timeIndex = Math.min(index, 48);
  const hour: number = Math.floor(timeIndex / 2);
  const minute: string = timeIndex % 2 === 0 ? '00' : '30';
  return (hour < 10 ? `0${hour}` : hour.toString()) + minute;
}

export function formatHour(hour: number): string {
  if (hour === 12) return '12 noon';
  if (hour === 0 || hour === 24) return '12 midnight';
  if (hour < 12) return `${hour}am`;
  return `${hour - 12}pm`;
}

export function getTimeAsDate(time: string | number, date: Date = new Date()): Date {
  const dateNumber = typeof time === 'string' ? parseInt(time, 10) : time;
  return setHours(setMinutes(startOfDay(date), dateNumber % 100), Math.floor(dateNumber / 100));
}

export function formatTime(time: string | number): string {
  const timeNumber = typeof time === 'string' ? parseInt(time, 10) : time;

  if (timeNumber === 0) return '12 midnight';
  if (timeNumber === 1200) return '12 noon';

  return format(getTimeAsDate(timeNumber), 'h:mm a').toLowerCase();
}

// Create a new date object with time from the second date object
export function setTime(date: Date, time: Date): Date {
  return setHours(setMinutes(setSeconds(date, getSeconds(time)), getMinutes(time)), getHours(time));
}

export const SCHOOLDAYS: DayText[] = [
  'Monday',
  'Tuesday',
  'Wednesday',
  'Thursday',
  'Friday',
  'Saturday',
];
export const DEFAULT_EARLIEST_TIME: LessonTime = '1000';
export const DEFAULT_LATEST_TIME: LessonTime = '1800';

// Given an array of lessons, we calculate the earliest and latest timings based on the lesson timings.
// This bounds will then be used to decide the starting and ending hours of the timetable.
export function calculateBorderTimings(
  lessons: Lesson[],
  period?: TimePeriod,
): { startingIndex: number; endingIndex: number } {
  let earliestTime: number = convertTimeToIndex(DEFAULT_EARLIEST_TIME);
  let latestTime: number = convertTimeToIndex(DEFAULT_LATEST_TIME);
  lessons.forEach((lesson) => {
    earliestTime = Math.min(earliestTime, convertTimeToIndex(lesson.startTime));
    latestTime = Math.max(latestTime, convertTimeToIndex(lesson.endTime));
  });

  // Consider time range of period, if applicable
  if (period != null) {
    earliestTime = Math.min(earliestTime, convertTimeToIndex(period.startTime));
    latestTime = Math.max(latestTime, convertTimeToIndex(period.endTime));
  }

  return {
    startingIndex: earliestTime % 2 === 0 ? earliestTime : earliestTime - 1, // floor to earliest hour.
    endingIndex: latestTime % 2 === 0 ? latestTime : latestTime + 1, // ceil to latest hour.
  };
}

/**
 * Gets the current time in hours, 0915 -> 9, 1315 -> 13
 * @deprecated Use date injected by withTimer instead
 */
export function getCurrentHours(
  now: Date = new Date(), // Used for tests only
): number {
  return now.getHours();
}

/**
 * Gets the current time in hours, 0915 -> 15, 1345 -> 45
 * Current time to always match Singapore's
 *
 * @deprecated Use date injected by withTimer instead
 */
export function getCurrentMinutes(
  now: Date = new Date(), // Used for tests only
): number {
  return now.getMinutes();
}

// Monday = 0, Friday = 4, Sunday = 6
export function getDayIndex(date: Date = new Date()): number {
  return getISODay(date) - 1;
}

/**
 * Return a copy of the original Date incremented by the given number of days
 *
 * @deprecated Use addDays from date-fns
 */
export function daysAfter(startDate: Date, days: number): Date {
  const d = new Date(startDate.valueOf());
  d.setUTCDate(d.getUTCDate() + days);
  return d;
}

/**
 * Converts a Date object representing an event happening in Singapore time
 * and outputs a new Date object with the local time in SGT. This is useful
 * in conjunction with format from date-fns since it always use local time when
 * formatting output.
 *
 * @example
 *     // Exam is at 9AM 23rd of October 2016
 *     const examDate = new Date('2016-11-23T01:00:00.000Z');
 *     format(examDate, 'dd-MM-yyyy p');
 *     // => "23-11-2016 9:00 AM", no matter where the user machine's TZ is
 */
export function toSingaporeTime(date: string | number | Date): Date {
  const localDate = new Date(date);
  return new Date(localDate.getTime() + (localDate.getTimezoneOffset() - SGT_OFFSET) * 60 * 1000);
}

/**
 * Convert an ISO date string, eg. 2018-10-12 to a Date object with the
 * given date and time set to midnight SGT (UTC+8)
 */
export function parseDate(string: string): Date {
  return parseISO(`${string}T00:00+0800`);
}
