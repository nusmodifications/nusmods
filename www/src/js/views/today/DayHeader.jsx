// @flow

import React from 'react';
import { castArray, last } from 'lodash';
import { format } from 'date-fns';

import getWeatherIcon from 'views/components/icons/weather';
import Tooltip from 'views/components/Tooltip';
import styles from './DayHeader.scss';

type Props = {|
  +date: Date | Date[],
  +forecast?: ?string,
  +offset: number, // number of days from today
|};

export function HeaderDate({ children, offset }: {| +children: Date, +offset: number |}) {
  let title;
  let subtitle;

  if (offset < 2) {
    title = offset === 0 ? 'Today' : 'Tomorrow';
    subtitle = format(children, 'iiii, do MMMM');
  } else {
    title = format(children, 'iiii');
    subtitle = format(children, 'do MMMM');
  }

  return (
    <time dateTime={children.toISOString()}>
      <span className={styles.date}>{subtitle}</span> {title}
    </time>
  );
}

export default function(props: Props) {
  const Icon = props.forecast ? getWeatherIcon(props.forecast) : null;

  const dates = castArray(props.date);

  return (
    <header className={styles.header}>
      <h2>
        <div>
          <HeaderDate offset={props.offset}>{dates[0]}</HeaderDate>
        </div>
        {dates.length > 1 && (
          <>
            <div className={styles.to}> to </div>
            <div>
              <HeaderDate offset={props.offset + dates.length - 1}>{last(dates)}</HeaderDate>
            </div>
          </>
        )}
      </h2>

      {Icon && (
        <Tooltip content={props.forecast} placement="bottom" offset={0}>
          <div className={styles.weather} aria-label={props.forecast}>
            <Icon />
          </div>
        </Tooltip>
      )}
    </header>
  );
}
