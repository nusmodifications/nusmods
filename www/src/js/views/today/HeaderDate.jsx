// @flow

import React from 'react';
import { format } from 'date-fns';
import styles from './TodayContainer.scss';

// Get the subtitle and title headings
type Props = {|
  +children: Date,
  +offset: number,
|};

export default function HeaderDate({ children, offset }: Props) {
  let title;
  let subtitle;

  if (offset < 2) {
    title = offset === 0 ? 'Today' : 'Tomorrow';
    subtitle = format(children, 'iiii, do MMMM');
  } else {
    title = format(children, 'iiii');
    subtitle = format(children, 'do MMMM');
  }

  return (
    <time dateTime={children.toISOString()}>
      <span className={styles.date}>{subtitle}</span>
      {title}
    </time>
  );
}
