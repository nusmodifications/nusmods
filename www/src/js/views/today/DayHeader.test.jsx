// @flow

import React from 'react';
import { shallow } from 'enzyme';
import { daysAfter } from 'utils/timify';
import DayHeader from './DayHeader';
import HeaderDate from './HeaderDate';
import styles from './DayHeader.scss';

describe(DayHeader, () => {
  const today = new Date('2016-11-23T09:00+0800');

  test('render one date when one date is specified', () => {
    expect(shallow(<DayHeader date={today} offset={0} />).find(HeaderDate)).toHaveLength(1);
    expect(shallow(<DayHeader date={[today]} offset={0} />).find(HeaderDate)).toHaveLength(1);
  });

  test('render two date when more than one date is specified', () => {
    const wrapperOne = shallow(<DayHeader date={[today, daysAfter(today, 1)]} offset={0} />);

    const wrapperTwo = shallow(
      <DayHeader date={[today, daysAfter(today, 1), daysAfter(today, 2)]} offset={0} />,
    );

    expect(wrapperOne.find(HeaderDate)).toHaveLength(2);
    expect(wrapperTwo.find(HeaderDate)).toHaveLength(2);
  });

  test('render weather when it is specified', () => {
    const wrapper = shallow(<DayHeader date={today} offset={0} forecast="Cloudy" />);

    expect(wrapper.find(`.${styles.weather}`).exists()).toBe(true);
  });
});
