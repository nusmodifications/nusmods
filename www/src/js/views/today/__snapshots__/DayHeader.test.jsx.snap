// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderDate render date as day of week if offset more than one 1`] = `
<time
  dateTime="2016-11-23T01:00:00.000Z"
>
  <span
    className="date"
  >
    23rd November
  </span>
   
  Wednesday
</time>
`;

exports[`HeaderDate render date as day of week if offset more than one 2`] = `
<time
  dateTime="2016-11-23T01:00:00.000Z"
>
  <span
    className="date"
  >
    23rd November
  </span>
   
  Wednesday
</time>
`;
